packages/grpc-js/src/server.ts

```diff
@@ -1421,41 +1420,72 @@ export class Server {
       }
 
       if (this.keepaliveTimeMs < KEEPALIVE_MAX_TIME_MS) {
-        keeapliveTimeTimer = setInterval(() => {
-          keepaliveTimeoutTimer = setTimeout(() => {
-            sessionClosedByServer = true;
-            session.close();
+        keepaliveInterval = setInterval(() => {
+          const keepaliveTimeout = setTimeout(() => {
+            if (keepaliveInterval) {
+              clearInterval(keepaliveInterval);
+              keepaliveInterval = null;
+              sessionClosedByServer = true;
+              this.trace('Connection dropped by keepalive timeout');
+              session.close();
+            }
           }, this.keepaliveTimeoutMs);
-          keepaliveTimeoutTimer.unref?.();
+          keepaliveTimeout.unref?.();
 
           try {
-            session.ping(
-              (err: Error | null, duration: number, payload: Buffer) => {
-                if (keepaliveTimeoutTimer) {
-                  clearTimeout(keepaliveTimeoutTimer);
+            if (
+              !session.ping(
+                (err: Error | null, duration: number, payload: Buffer) => {
+                  clearTimeout(keepaliveTimeout);
+                  if (err) {
+                    if (keepaliveInterval) {
+                      clearInterval(keepaliveInterval);
+                      keepaliveInterval = null;
+                    }
+                    sessionClosedByServer = true;
+                    this.trace(
+                      'Connection dropped due to error with ping frame ' +
+                        err.message +
+                        ' return in ' +
+                        duration
+                    );
+                    session.close();
+                  }
                 }
-
-                if (err) {
-                  sessionClosedByServer = true;
-                  this.trace(
-                    'Connection dropped due to error of a ping frame ' +
-                      err.message +
-                      ' return in ' +
-                      duration
-                  );
-                  session.close();
-                }
-              }
-            );
+              )
+            ) {
+              throw new Error('Server keepalive ping send failed');
```

Comment:
I don't like the `throw` immediately into a `catch` block here. If you remove the cleanup stuff in the `catch` block, I think it's reasonable to just have separate `trace` and `session.destroy` calls here.

Reply:
#2139 reported that `session.ping` can throw errors.